resources/lib/library.py

```diff
@@ -1305,6 +1405,122 @@ def addons( self ):
         
         self.loadedAddOns = True
         return self.loadedAddOns
+
+    def hasPluginEntryPoint( self, path ):
+        # Check if an addon has a plugin entry point by parsing its addon.xml file
+        try:
+            tree = xmltree.parse( os.path.join( path, "addon.xml" ) ).getroot()
+            for extension in tree.findall( "extension" ):
+                if "point" in extension.attrib and extension.attrib.get( "point" ) == "xbmc.python.pluginsource":
+                    # Find out what content type it provides
+                    provides = extension.find( "provides" )
+                    if provides is None:
+                        return None
+                    return provides.text.split( " " )
+
+
+        except:
+            return None
+        return None
+    
+    def detectPluginContent(self, item):
+        #based on the properties in the listitem we try to detect the content
+        
+        if not item.has_key("showtitle") and not item.has_key("artist"):
+            #these properties are only returned in the json response if we're looking at actual file content...
+            # if it's missing it means this is a main directory listing and no need to scan the underlying listitems.
+            return None
+
+        if not item.has_key("showtitle") and item.has_key("artist"):
+            ##### AUDIO ITEMS ####
+            if item["artist"][0] == item["title"]:
+                return "artists"
+            elif item["album"] == item["title"]:
+                return "albums"
+            elif (item["type"] == "song" or (item["artist"] and item["album"])):
+                return "songs"
+        else:    
+            ##### VIDEO ITEMS ####
+            if (item["showtitle"] and not item["artist"]):
+                #this is a tvshow, episode or season...
+                if (item["season"] > -1 and item["episode"] == -1):
+                    return "seasons"
+                elif item["season"] > -1 and item["episode"] > -1:
+                    return "episodes"
+                else:
+                    return "tvshows"
+            elif (item["artist"]):
+                #this is a musicvideo!
+                return "musicvideos"
+            elif (item["imdbnumber"] or item["mpaa"] or item["trailer"] or item["studio"]):
+                return "movies"
+
+        return None
+
+    def widgets( self ):
+        if self.loadedWidgets == True:
```

Comment:
since "True" and "False" are singletons it is recommended to use "is True" or just "if self.loadedWidgets:"
see here for example for bit more details: https://stackoverflow.com/questions/27276610/boolean-identity-true-vs-is-true